### ISSUE TITLE
Impossible to get running OPX in GNS3

### ISSUE THREAD

**Issue by q2dg** [open-switch/opx-docs#13]
https://github.com/open-switch/opx-docs/wiki/Run-virtual-machine explains a little bit how to set up OPX in VirtualBox. Apart from appearing a "dummy0" and a "npu-0" interface in my VM which aren't explained (and I don't know for what they worth),  my main concern is about how to use e101-NNN-x interfaces as a simple switch ports in a GNS3 environment. This software expects interfaces to connect should be in VirtualBox's "Not connected" mode to be able to manage them but VirtualBox by default only shows 4 interfaces, so with lo, eth0, dummy0 and nps-0, the quota has already been filled!!

In other words: how can I get rid of dummy-0 and npu-0 interfaces so, at least, I can have two "e101-NNN-x" interfaces recognized by VirtualBox to use as a regular (although only 2-port) switch inside GNS3 labs?

In other words, why there are so may "e101-NNN-x" interfaces if VirtualBox can't recognize them (and, therefore, GNS can't handle them)?

Thanks a lot

**Comment by mikelazar** [comment]
You do not need to get rid of dummy-0 and npu-0. You do not want to get rid of lo, it's the Linux loopback interface, and it's always there when networking is enabled in Linux.

As mentioned in the wiki, section "Setup VM network adapters":  'eth0' corresponds to the first adapter attached to the VM, e101-001-0 to the second adapter and so on, and e101-00N-1 to adapter N+1.

'lo' is the loopback interface - an internal "logical" network interface. In Linux, it will never correspond to an Ethernet device.

'npu-0' and 'dummy-0' are also logical interfaces created for OPX - but they will not correspond to Ethernet devices (virtual network adapters)

Please make sure that you use Intel PRO/1000 MT Desktop (82540EM) as the type of all network adapters in Virtual Box. This needs to be the setup **before** you install OPX. Otherwise, the Linux PCI device discovery does not work properly - it enumerates Ethernet devices in the wrong order and you may end up with eth0 assigned to a random adapter rather than adapter 1.

Unfortunately, Virtual Box only displays four virtual adapters in the GUI. It actually supports 8 adapters, but they can only be managed through the Virtual Box command line. GNS3 might be able to make use of them anyway.

As far as I recall GNS3 behavior with VirtualBox, you need to create a link between two adapters on different VMs for the adapters to have the "Cable Connected" checkmark set. That way GNS3 simulates the presence of a physical cable.

I would suggest that you experiment with Virtual Box only before attempting to use GNS3 with VirtualBox.

**Comment by q2dg** [comment]
Thanks a lot for your great explanation!!

What I wanted to know, in summary, is if there is some way to explicitely bind an e101-xxx-x interface to a specific "tab" interface in VirtualBox.

If you simply set a "tab" interface in VirtualBox enabled but in "Not connected" state, GNS3 is able to use it right away.

Thanks again

**Comment by q2dg** [comment]
Well, I don't know why it is closed. I've build a VirtualBox OPX VM from your official Vagrant files and I have the problem of linking "software" e101-xxx-x" interfaces to "real" tab VirtualBox ones. If it doesn't work, "e101-xxx-x" interfaces are useless in a virtualized environment. Hence this issue.

**Comment by q2dg** [comment]
Apparently, yes. I have a bunch of e101-xxx-x interfaces (more than 30!) in addition to eth0 (that works perfectly), dummy0 and npu-0 ones but e101-xxx-x aren't tied to "tab" interfaces in VirtualBox (or, at least, I haven't found the logic). 

Maybe the key is in the first point: "Use N+1 Virtual Network Adapters for the VM, N for the simulated NPU ports **(but in which mode? bridged? host-only? NAT?...)** and one for the management port (eth0). 

I've used 2.3.0 vagrant files from https://app.vagrantup.com/opx/boxes/2.3.0 and VirtualBox 5.2.8 (in a Fedora 27 system and a x86_64 4.15.17-300 kernel).

Thanks a lot for your patience.

**Comment by mikelazar** [comment]
From the perspective of the of whether Ethernet interfaces are discovered or not, the mode (bridged, host-only makes no difference). 

When you use standalone VirtualBox VMs (**NOT in GNS3**), we recommend "internal network" to configure a point-to-point link between Ethernet devices on two different VMs. You should use the same internal network name at both end-points. For more information, please refer to VirtualBox documentation. Please note that if you use VirtualBox/GNS3, point to point links between VMs are created using a custom GNS3 driver.

You will always see a large number (32, by default) of e101-00N-0 interface. These are just 'tap' devices that on 'real' HW correspond to ASIC/switch ports (see  _**ip tuntap**_) . In a VM, we associate (some of) these tuntap interfaces to Ethernet devices that are moved to a separate namespace at startup. These Ethernet interfaces simulate front panel/ASIC ports. The NAS packet I/O sends traffic back and forth between the e101-00N-0 devices and the the ASIC ports - in a real switch. On a VM, we use the Ethernet interfaces that correspond to the virtual network adapter.

How to debug (I assume that you get network connectivity through eth0 - NAT or something else):

install pciutils (unfortunately, this package is not installed by default)

sudo apt-get update
sudo apt-get install pciutils

sudo lspci -v | grep Ethernet 
should show as many Ethernet interfaces as network adapters you have configured:

00:04.0 Ethernet controller: Intel Corporation 82540EM Gigabit Ethernet Controller (rev 03)
00:05.0 Ethernet controller: Intel Corporation 82540EM Gigabit Ethernet Controller (rev 03)
00:06.0 Ethernet controller: Intel Corporation 82540EM Gigabit Ethernet Controller (rev 03)
00:07.0 Ethernet controller: Intel Corporation 82540EM Gigabit Ethernet Controller (rev 03)
...

Again, please remember that all adapters need to be set to Intel Pro Desktop 82540

Display the "switch (virtual) ports".

# sudo ip netns  list
vportnetns

# sudo ip netns exec vportnetns ip link show
...
7: vport3: <BROADCAST,MULTICAST,PROMISC,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP mode DEFAULT group default qlen 1000
    link/ether 00:6c:d4:c0:f9:09 brd ff:ff:ff:ff:ff:ff
8: vport1: <BROADCAST,MULTICAST,PROMISC,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP mode DEFAULT group default qlen 1000
    link/ether 00:6c:d4:c0:f9:01 brd ff:ff:ff:ff:ff:ff
9: vport2: <BROADCAST,MULTICAST,PROMISC,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP mode DEFAULT group default qlen 1000
    link/ether 00:6c:d4:c0:f9:05 brd ff:ff:ff:ff:ff:ff

You should have N-1 vports - starting from vport1, where is the number of network adapters in VirtualBox

vport1 is associated to vbox adapter 2 (tap device e101-001-0)
vport2 is associated to vbox adapter 3 (tap device e101-002-0)
vport3 is associated to vbox adapter 4 (tap device e101-003-0)
...

As is the case for real HW, the e101-00N-0 operational state follow the operational state of the vport devices (which simulate HW ports for the VM).

List VM interfaces for your VM - use VBoxManage - from the "host"

# VBoxManage showvminfo YourVMName | grep NIC
...
NIC 4:           MAC: 080027410A69, Attachment: Internal Network 'intnetLink1', Cable connected: on, Trace: off (file: none), Type: 82540EM, Reported speed: 0 Mbps, Boot priority: 0, Promisc Policy: allow-all, Bandwidth group: none
...

Please remember to set the Promisc Policy to "allow-all". We change the MAC addresses of the vports - so they do not match the MAC address on the host side. "Allow-all" allows traffic to be delivered to the VM.

If you still have troubles, please post the output of these commands.

**Comment by jeff-yin** [comment]
@q2dg how's it going? did the above help?

**Comment by q2dg** [comment]
@jeff-yin Sorry for the delay. This week-end I try and inform you. Thanks a lot!

**Comment by jeff-yin** [comment]
@q2dg any update?

**Comment by jeff-yin** [comment]
Closing due to lack of update

**Comment by q2dg** [comment]
Sorry for not replying before, I've been really sick last month.

Anyway, I've tried the commands you kindly explained above (your splendid explanation could be added to wiki!) and they work. But I'm insisting in the use of GNS3 because if VMs are in "internal network", they don't need a switch at all because they all are connected between themselves already! 

What I want to do is a graphical design of a network map with VMs linked through Open-switch switches in a flexible way. And GNS3 "forces" to have the VM's network interfaces in "not managed" mode to be able to manage them instead.

But it's the same...if it (still) can't be possible, no problem.
Thanks a lot for your patience and help.

PD: Before _sudo ip netns exec vportnetns ip link show_ command I had to execute _sudo ip netns exec vportnetns ip link set up dev vportX_ command but don't make me much attention.
